ffi/src/lib.rs

```diff
@@ -113,7 +114,7 @@ impl TryFromStringSlice for String {
     ///
     /// The slice must be a valid (non-null) pointer, and must point to the indicated number of
     /// valid utf8 bytes.
-    unsafe fn try_from_slice(slice: KernelStringSlice) -> DeltaResult<String> {
+    unsafe fn try_from_slice(slice: &KernelStringSlice) -> DeltaResult<String> {
```

Comment:
@scovich this is obviously a significant change. I can't think of a way that it causes an issue though, since we basically only require the lifetime of the slice pointer to be "this function", which was the same as before. We call `slice.into()` which allocates, so as soon as this returns, the pointer can become invalid and that's fine.

This is more flexible because we can have a struct with a KernelStringSlice, and then can turn it into a String, without having to consume the whole struct.

Reply:
Quick drive-by response (will try to look more later): How often do we need to consume a kernel string slice from a struct without consuming the struct? And in such cases, should we just create a new slice object since it's anyway just a borrowed reference to a(n externally owned) string?

Currently, `KernelStringSlice` is neither `Copy` nor `Clone`, in an attempt to encourage at least semi-sane semantics. However, it can still be "moved" from, which includes returning it from a function or assigning it to a heap-allocated struct field, so I'm not sure how much protection we _really_ get? It seems like any copy or clone can only outlive the original if it gets moved-from -- e.g. returned from a function -- but the same is already true for the original instance?? Maybe this is another case where the compiler can't save us and we have to rely on idiomatic code patterns instead?

Reply:
However, IIRC we moved to passing by value instead of by reference, because it was causing really weird lifetime issues. So I'd be very cautious moving back to references without understanding that previous issue better. I don't remember details any more, sadly...

Reply:
Update: Thinking more, the lifetime safety issues mentioned above seem orthogonal to whether `try_from_slice` takes a value or a reference? And if we look at it in isolation, taking a reference is actually the right approach, because creating a `String` really doesn't consume the backing string the slice references. Similar to how the language has [impl From<&str> for String](https://doc.rust-lang.org/std/string/struct.String.html#impl-From%3C%26str%3E-for-String)?

Reply:
Yeah, I think that captures nicely what I was more vaguely thinking. Basically, because `try_from_slice` doesn't actually consume the slice, it makes sense to take it as a reference.

I've added to the comment of this function that it doesn't consume the slice. LMK if you have further concerns about this change though.